pkg/middleware/openapi/openapi.go

```diff
@@ -157,59 +198,119 @@ func (v *Validator) validateResponse(w *bufferingResponseWriter, r *http.Request
+// that actually insigated the request so it can be propagated to and used
+// by any service.  This is called only by other system services as
+// identified by the use of mTLS.
+func extractPrincipal(ctx context.Context, r *http.Request) (context.Context, error) {
+	data := r.Header.Get(principal.Header)
+	if data == "" {
+		return nil, fmt.Errorf("%w: principal header not present", ErrHeader)
+	}
+
+	// Use the certificate of the service that actually called us.
+	// The one in the context is used to propagate token binding information.
+	certRaw, err := util.GetClientCertificateHeader(r.Header)
 	if err != nil {
-		errors.HandleError(w, r, errors.OAuth2ServerError("route lookup failure").WithError(err))
-		return
+		return nil, err
 	}
 
-	// Propagate the client certificate now so it's available in the request validation
-	// in case its required for a bound access token.
-	ctx, err := authorization.ExtractClientCert(r.Context(), r.Header)
+	certificate, err := util.GetClientCertificate(certRaw)
 	if err != nil {
-		errors.HandleError(w, r, errors.OAuth2InvalidRequest("certificate propagation failure").WithError(err))
+		return nil, err
+	}
 
-		return
+	p := &principal.Principal{}
+
+	if err := client.VerifyAndDecode(p, data, certificate); err != nil {
+		return nil, err
 	}
 
-	// Make a shallow copy of the request with the new context.  OpenAPI validation
-	// will read the body, and replace it with a new buffer, so be sure to use this
-	// version from here on.
-	r = r.WithContext(ctx)
+	return principal.NewContext(ctx, p), nil
+}
+
+// extractOrGeneratePrincipal extracts the principal if mTLS is in use, for service to service
+// API calls, otherwise it generates it from the available information.
+func (v *Validator) extractOrGeneratePrincipal(ctx context.Context, r *http.Request, params map[string]string) (context.Context, error) {
+	if util.HasClientCeritifcateHeader(r.Header) {
+		newCtx, err := extractPrincipal(ctx, r)
+		if err != nil {
+			return nil, err
+		}
+
+		return newCtx, nil
+	}
+
+	return v.generatePrincipal(ctx, params), nil
+}
+
+// validateAndAuthorize performs OpenAPI schema validation of the request, and also
+// triggers an authentication callback when the APi is marked as requiring it.
+// This will read the request body from the original and replace it with a buffer.
+// As we are doing a shallow copy to inject authentication context information you
+// must use the returned request for the HTTP handlers.
+func (v *Validator) validateAndAuthorize(ctx context.Context, r *http.Request, route *routers.Route, params map[string]string) (*http.Request, *openapi3filter.ResponseValidationInput, error) {
+	// If mTLS is in use, then the access token *may* be bound to the X.509 private key,
+	// but only in the case where a service is using a client credentials grant.
+	// As all services act on behalf of clients, we only want the client certificate to
```

Comment:
```suggestion
	// As all services act on behalf of clients, we want only the client certificate to
```

Is this change correct, as opposed to 

> As all services act on behalf of clients, we want the client certificate to
> be propagated **only** to the identity service during authentication (userinfo call) and
> authorization (ACL call)

Reply:
OK; I ask because the request with that context is returned and used in subsequent calls. So if the certificate is only wanted for these calls, that context should be discarded afterward.